### PR TITLE
fix: catch uncaught exception from append loop and add expoential retry to reconnection 

### PR DESCRIPTION
**Pull request by GaoleMeng**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-bigquerystorage/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> ☕️

If you write sample code, please follow the [samples format](
https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md).

### CODE REVIEW

File: google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/ConnectionWorker.java [hunk 1]
```diff
@@ -196,6 +196,12 @@ class ConnectionWorker implements AutoCloseable {
    */
   private final String writerId = UUID.randomUUID().toString();
 
+  /*
+   * Test only exception behavior testing params.
+   */
+  private RuntimeException testOnlyRunTimeExceptionInAppendLoop;
```
Comment (yirutang) [inline google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/ConnectionWorker.java:202]: init this to null?

File: google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/ConnectionWorker.java [hunk 2]
```diff
@@ -383,6 +410,22 @@ private void maybeWaitForInflightQuota() {
     inflightWaitSec.set((System.currentTimeMillis() - start_time) / 1000);
   }
 
+  @VisibleForTesting
+  static long calculateSleepTime(long retryCount) {
+    return (long) Math.pow(5, retryCount);
```
Comment (yirutang) [inline google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/ConnectionWorker.java:415]: Are we backing off too much? Maybe base 2 with a max interval of say 1 min.

File: google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/ConnectionWorker.java (hunk 1 above)
Comment (GaoleMeng) [reply google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/ConnectionWorker.java]: Done

File: google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/ConnectionWorker.java (hunk 2 above)
Comment (GaoleMeng) [reply google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/ConnectionWorker.java]: added a max, but it's millisecond, do we still think 5 is too much?

Comment (yirutang) [reply google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/ConnectionWorker.java]: We have a total time out of retry 5 minutes. So we end up can only retry 7 times? As long as there is a max, I am OK.

Comment (GaoleMeng) [reply google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/ConnectionWorker.java]: Let's change to 2 to retry more times

**Review by yirutang — APPROVED** [review]

**Pull request closed without merging**